### PR TITLE
Add inceptionv4 pretrained model

### PR DESCRIPTION
**Pull request by Cadene**
Made with [Tensorflow Model Zoo for Torch7 and PyTorch](https://github.com/Cadene/tensorflow-model-zoo.torch)

### CODE REVIEW

**Comment by soumith** [comment]
is it the same input normalization as all other models in the current zoo?

**Comment by Cadene** [comment]
I am not sure, but it might be [this](https://github.com/tensorflow/models/blob/master/slim/preprocessing/inception_preprocessing.py#L273) i.e. sub(0.5) then mul(2.0)

**Comment by apaszke** [comment]
We could probably adjust the first layer weights to match the `torchvision` style normalization.

**Comment by Cadene** [comment]
We could also add the normalization information [that way](https://github.com/Cadene/torchnet-vision/blob/master/models/inceptionv3.lua#L55), i.e. not as a layer, but as an attribute.

**Comment by Cadene** [comment]
I used the [imagenet example](https://github.com/pytorch/examples/tree/master/imagenet) to evaluate the model.
I've got  80.062 % accuracy top1 on valset and 94.926 accuracy top 5.
I am currently downloading the testset.

Beware, the targets are associated to the labels in the same order, but inceptionv4 has 1001 classes. The first class (index 0) is said to be the "dummy class". Thus, I shifted the targets number that way `target = target.cuda(async=True).add(1)`

I preprocessed images that way:
```python
normalize = transforms.Normalize(mean=[0.5, 0.5, 0.5],
                                 std=[0.5, 0.5, 0.5])
...
val_loader = torch.utils.data.DataLoader(
        datasets.ImageFolder(valdir, transforms.Compose([
            transforms.Scale(342),
            transforms.CenterCrop(299),
            transforms.ToTensor(),
            normalize
        ])),
        batch_size=args.batch_size, shuffle=False,
        num_workers=args.workers, pin_memory=True)
```

**Comment by Cadene** [comment]
@apaszke I am not sure, but I don't think it's possible, because you must add a bias by color channel (the normalization in pytorch vision is made by channel).

The problem is the following.
In Tensorflow, the input image is normalized by mean=0.5, std=0.5.
In PyTorch/Vision, the input image is normalized by channel mean=0.485, 0.456, 0.406, std=0.229, 0.224, 0.225.
How to adjust the pretrained weights of the first convolutional layers to match the Pytorch normalization ? 
Not a difficult problem, but here we have a strong constraint. We can't use a bias per color channel, because the conv2d has only one bias. 

Let's simplify the equation. We have x one pixel, w one weight, y the output.
(x - 0.5) / 0.5 * w = y

Without biases, we have to find an epsilon e such as x, w and y remain unchanged.
(x - 0.485 ) / 0.229 * (w * e) = y
(x - 0.485 ) / 0.229 * (w * e) = (x - 0.5) / 0.5 * w

(x - 0.485 ) / 0.229 * e = (x - 0.5) / 0.5
e = (x - 0.5) / 0.5 * 0.229 / (x - 0.485)

Thus, without biases per color channel you can't do it, because the epsilon depends on the input x.

**Comment by Cadene** [comment]
If it is not possible, I think we should associate pretrained parameters to their normalization. 

Let me explain myself. I saw that you retrained VGG on ImageNet. That's really cool and we could do the same with InceptionV4, but it would be nice also that for instance people who want to reproduce results from papers using the VGG pretrained with caffe (with a different kind of normalization maybe) should be able to download the pretrained parameters of their choice associated with an information about how to normalize the input.
